llvm/lib/Transforms/Utils/LoopRotationUtils.cpp

```diff
@@ -551,7 +606,17 @@ bool LoopRotate::rotateLoop(Loop *L, bool SimplifiedLatch) {
       if (L->hasLoopInvariantOperands(Inst) && !Inst->mayReadFromMemory() &&
           !Inst->mayWriteToMemory() && !Inst->isTerminator() &&
           !isa<DbgInfoIntrinsic>(Inst) && !isa<AllocaInst>(Inst)) {
+
+        if (LoopEntryBranch->getParent()->IsNewDbgInfoFormat) {
+          auto DbgValueRange =
+              LoopEntryBranch->cloneDebugInfoFrom(Inst, NextDbgInst);
```

Comment:
I'm confused - it looks like `NextDbgInst` is only ever `std::nullopt` or `I->getDbgValueRange().begin();`. Looking at the `cloneDebugInfoFrom` implementation in main, wouldn't those inputs cause the same behaviour? What am I missing?

Reply:
Even more extra commentary added in the latest commit, hopefully that illustrates a bit better? Not that it helps, but this is certainly the dumbest part of all RemoveDIs -- it's also unavoidable given that we're separating the iterator/cursor of Real Instructions (TM) from the positions of debug-info. Hopefully in the future we'll refactor this into a utility that takes care of it.

Reply:
That makes sense now, thanks for bearing with me!